src/data_types/indexed_tensor.hpp

```diff
@@ -0,0 +1,209 @@
+ * This tool is used in the function index and shouldn't be called by users directly.
+ * @param tensor The tensor that is being indexed.
+ * @tparam TypeSeqCharIds A type sequence of characters (saved as integral_constants) describing the
+ *                      index pattern used to index the tensor.
+ * @return An IndexedTensor object.
+ */
+template <class TypeSeqCharIds, class TensorType, std::size_t... I>
+auto internal_index(TensorType& tensor, std::index_sequence<I...>)
+{
+    return IndexedTensor<
+            TensorType,
+            ddc::detail::TypeSeq<VectorIndexIdMap<
+                    ddc::type_seq_element_t<I, TypeSeqCharIds>::value,
+                    typename TensorType::vector_index_set_t<I>>...>>(tensor);
+}
+
+template <
+        class GlobalTensorIndexIdMap,
+        std::size_t Is,
+        class ResultIndexedTensorType,
+        class... IndexedTensorType>
+void internal_tensor_mul_elem(ResultIndexedTensorType& result, IndexedTensorType... t)
+{
+    using TensorTuple = std::tuple<typename IndexedTensorType::tensor_type...>;
+    using ElementType = typename std::tuple_element_t<0, TensorTuple>::element_type;
+    using Idx = get_nth_tensor_index_element_from_map_t<Is, GlobalTensorIndexIdMap>;
+    if constexpr (std::is_same_v<ResultIndexedTensorType, ElementType>) {
+        result
+                += ((t().template get<extract_sub_tensor_element_t<
+                             GlobalTensorIndexIdMap,
+                             typename IndexedTensorType::index_pattern,
+                             Idx>>())
+                    * ...);
+    } else {
+        result().template get<extract_sub_tensor_element_t<
+                GlobalTensorIndexIdMap,
+                typename ResultIndexedTensorType::index_pattern,
+                Idx>>()
+                += (t().template get<extract_sub_tensor_element_t<
+                            GlobalTensorIndexIdMap,
+                            typename IndexedTensorType::index_pattern,
+                            Idx>>()
+                    * ...);
+    }
+}
+
+template <
+        class GlobalTensorIndexIdMap,
+        class ResultIndexedTensorType,
+        class... IndexedTensorType,
+        std::size_t... Is>
+void internal_tensor_mul(
+        ResultIndexedTensorType& result,
+        std::index_sequence<Is...>,
+        IndexedTensorType... t)
+{
+    ((internal_tensor_mul_elem<GlobalTensorIndexIdMap, Is>(result, t...)), ...);
+}
+
+} // namespace details
+
+} // namespace tensor_tools
+
+/**
+ * @brief A tool to build an IndexedTensor object from a set of indices.
+ * This object can be used to carry out a tensor multiplication using the tensor_mul function.
+ * Example:
+ * @code
+ * index<'i', 'j', 'k'>(my_3d_tensor)
+ * @endcode
+ * @param tensor The tensor that is being indexed.
+ * @tparam ids The characters describing the index pattern used to index the tensor.
+ * @return An IndexedTensor object.
+ */
+template <char... ids, class TensorType>
```

Comment:
Should we check the dimensions match ? Or is it done somewhere else ?

Reply:
You are right. I should check that the correct number of index ids are passed.

I'm not sure if this is what you meant though or if you wanted to check something else?

Reply:
Done [5c168e3](https://github.com/gyselax/gyselalibxx/pull/114/commits/5c168e3024f230e7563ebe57c2e6cd5bdb8ab592)

Reply:
Seems good